### ISSUE TITLE
Further project layout conventions.

### ISSUE THREAD

**Issue by yeban** [monicadragan/GeneValidator#61]
Follow up to #58. Regarding [`lib/enumberable.rb`](https://github.com/monicadragan/GeneValidator/blob/master/lib/genevalidator/enumerable.rb)

Convention is to put modifications to core or stdlib classes in `ext/`. And the following approach to extending a core class / module is preferred over opening classes / modules:

```
# lib/ext/enumerable.rb
module GeneValidator
  module Enumerable
    def sum
    end
    ...  
  end
end

module Enumerable
  include GeneValidator::Enumerable
end
```

The advantage of this approach is that if there's ever a need to know where did `Enumerable#sum` come from:

```
# assume an Enumerable in GV
a = [1, 2]
a.method(:sum).owner
=> GeneValidator::Enumerable
```

It's clear where the method comes from - `lib/ext/enumerable.rb`. Works even if you borrow 'extensions' from other gems.

But is not so clear in the current approach:

```
a = [1, 2]
a.method(:sum).owner
=> Enumerable
```

**Comment by IsmailM** [comment]
Hey Priyam,

I seem to have some problem in getting this to work and was wondering if you could guide me with this...

The below works in pry:
```
module Enumerable
  def sum
    inject(0) { |accum, i| accum + i }
  end
end

a = [1,2]
a.sum
#=> 3
```

But this doesn't

```
module GeneValidator
  module Enumerable
    def sum
      inject(0) { |accum, i| accum + i }
    end
  end
end

module Enumerable
  include GeneValidator::Enumerable
end

a = [1,2]
a.sum
#=> NoMethodError: undefined method `sum' for [1, 2]:Array
```

Many Thanks

**Comment by yeban** [comment]
Ah. I mostly wrote that from memory, only partially testing the details.

It seems that adding methods 2-levels up the module chain doesn't reflect in the class. For example,

```
module B; end

class C
  include B
end

module A
  def hi
    puts 'hi'
  end
end

module B
  include A
end

C.new.hi
```

I wasn't aware of this.

Why not extend `Array` instead of `Enumerable`? That works.

```
module GeneValidator
  module Array
    def sum
      inject(0) { |accum, i| accum + i }
    end
  end
end

class Array
  include GeneValidator::Array
end

a = [1,2]
puts a.sum
puts a.method(:sum).owner
```

**Comment by yeban** [comment]
And since you are dropping pre 2.0 support, you could consider using refinements. Works similarly, except the scope of monkey-patched class (`Array` in this example) will be limited.

Refinements are file-scoped in 2.0, iirc. That is, you will have to "refine" (monkey-patch) the class in every file you want to use the "refinements" (new methods) in. Refinements work the way it really should 2.1 onwards - refinements will be in effect throughout GeneValidator module (or whichever module you refine the class in). I didn't get a chance to use refinements yet, so I may have a few details wrong.

**Comment by IsmailM** [comment]
Fixed in PR #63
